src/blockMonitor/blockProcessor.ts

```diff
@@ -135,45 +145,49 @@ export class BlockProcessor<TBlock extends IBlockStub> extends StartStopService
 
     protected async startInternal(): Promise<void> {
         // Make sure the current head block is processed
-        const initialBlockNumber = await this.provider.getBlockNumber();
-
-        await this.processBlockNumber(initialBlockNumber);
-
+        const currentHead = (await this.store.getLatestHeadNumber()) || (await this.provider.getBlockNumber());
+        await this.processBlockNumber(currentHead);
         this.provider.on("block", this.processBlockNumber);
     }
 
     protected async stopInternal(): Promise<void> {
         this.provider.removeListener("block", this.processBlockNumber);
     }
 
+    /**
+     * Adds a new listener for new head events.
+     * @param listener the listener for new head event; it will be passed the emitted `TBlock`.
+     */
+    public addNewHeadListener(listener: NewHeadListener<TBlock>) {
+        this.newHeadListeners.push(listener);
+    }
+
+    /**
+     * Removes `listener` from the list of listeners for new head events.
+     */
+    public removeNewHeadListener(listener: NewHeadListener<TBlock>) {
+        const idx = this.newHeadListeners.findIndex(l => l === listener);
+        if (idx === -1) throw new ApplicationError("No such listener exists.");
+
+        this.newHeadListeners.splice(idx, 1);
+    }
+
     // updates the new head block in the cache and emits the appropriate events
-    private processNewHead(headBlock: Readonly<TBlock>) {
-        this.mBlockCache.setHead(headBlock.hash);
-
-        // only emit events after it's started
-        if (this.started) {
-            // Go through the ancestry, add any block that up until (but excluding) the last block
-            // we emitted as head. If we never find a last emitted block, we emit all the ancestors in cache
-            let nearestEmittedHeadInAncestry: Readonly<TBlock> | null = null;
-            const blocksToEmit = [];
-            for (const block of this.blockCache.ancestry(headBlock.hash)) {
-                if (this.emittedBlockHeads.has(block)) {
-                    nearestEmittedHeadInAncestry = block;
-                    break;
-                } else {
-                    blocksToEmit.unshift(block);
-                }
-            }
+    private async processNewHead(headBlock: Readonly<TBlock>) {
+        try {
+            this.mBlockCache.setHead(headBlock.hash);
 
-            // Emit all the blocks past the latest block in the ancestry that was emitted as head
-            // In case of re-orgs, some blocks might be re-emitted multiple times.
-            for (const block of blocksToEmit) {
-                this.emit(BlockProcessor.NEW_BLOCK_EVENT, block);
-            }
+            await this.store.setLatestHeadNumber(headBlock.number);
```

Comment:
Done.

Do you think we should add any handling for potenial errors thrown by the listeners?
Currently, the db write happens in the try block, perhaps it shoud be outside (or better in a separate try/catch block), so that errors in listeners are logged, but don't mess upd with the updates.